From 6cdee6a8db030ff560bf652d9db7e60b3b18a3f3 Mon Sep 17 00:00:00 2001
From: Sebanc <sebanc.xda@gmail.com>
Date: Wed, 9 Oct 2019 21:49:52 +0200
Subject: [PATCH] Microsoft Surface Go devices: Add support for tablet mode
 detection and allow hid buttons to trigger long press actions

---
 drivers/platform/x86/intel-hid.c | 32 +++++++++++++++++++++++++++-----
 1 file changed, 27 insertions(+), 5 deletions(-)

diff --git a/drivers/platform/x86/intel-hid.c b/drivers/platform/x86/intel-hid.c
index 3201a83..2617684 100644
--- a/drivers/platform/x86/intel-hid.c
+++ b/drivers/platform/x86/intel-hid.c
@@ -60,15 +60,19 @@ static const struct key_entry intel_hid_keymap[] = {
 /* 5 button array notification value. */
 static const struct key_entry intel_array_keymap[] = {
 	{ KE_KEY,    0xC2, { KEY_LEFTMETA } },                /* Press */
-	{ KE_IGNORE, 0xC3, { KEY_LEFTMETA } },                /* Release */
+	{ KE_KEY,    0xC3, { KEY_LEFTMETA } },                /* Release */
 	{ KE_KEY,    0xC4, { KEY_VOLUMEUP } },                /* Press */
-	{ KE_IGNORE, 0xC5, { KEY_VOLUMEUP } },                /* Release */
+	{ KE_KEY,    0xC5, { KEY_VOLUMEUP } },                /* Release */
 	{ KE_KEY,    0xC6, { KEY_VOLUMEDOWN } },              /* Press */
-	{ KE_IGNORE, 0xC7, { KEY_VOLUMEDOWN } },              /* Release */
+	{ KE_KEY,    0xC7, { KEY_VOLUMEDOWN } },              /* Release */
 	{ KE_KEY,    0xC8, { KEY_ROTATE_LOCK_TOGGLE } },      /* Press */
-	{ KE_IGNORE, 0xC9, { KEY_ROTATE_LOCK_TOGGLE } },      /* Release */
+	{ KE_KEY,    0xC9, { KEY_ROTATE_LOCK_TOGGLE } },      /* Release */
+	{ KE_SW,     0xCA, { .sw = { SW_DOCK, 1 } } },	      /* Docked */
+	{ KE_SW,     0xCB, { .sw = { SW_DOCK, 0 } } },	      /* Undocked */
+ 	{ KE_SW,     0xCC, { .sw = { SW_TABLET_MODE, 1 } } }, /* Tablet */
+ 	{ KE_SW,     0xCD, { .sw = { SW_TABLET_MODE, 0 } } }, /* Clamshell */
 	{ KE_KEY,    0xCE, { KEY_POWER } },                   /* Press */
-	{ KE_IGNORE, 0xCF, { KEY_POWER } },                   /* Release */
+	{ KE_KEY,    0xCF, { KEY_POWER } },                   /* Release */
 	{ KE_END },
 };
 
@@ -343,6 +347,7 @@ static void notify_handler(acpi_handle handle, u32 event, void *context)
 	struct platform_device *device = context;
 	struct intel_hid_priv *priv = dev_get_drvdata(&device->dev);
 	unsigned long long ev_index;
+	struct key_entry *key;
 
 	if (priv->wakeup_mode) {
 		/*
@@ -389,6 +394,23 @@ wakeup:
 		}
 	}
 
+	if (priv->array) {
+		switch(event) {
+			case 0xc2: case 0xc4: case 0xc6: case 0xc8: case 0xce:
+				key = sparse_keymap_entry_from_scancode(
+					priv->array, event);
+				input_report_key(priv->array, key->keycode, 1);
+				input_sync(priv->array);
+				return;
+			case 0xc3: case 0xc5: case 0xc7: case 0xc9: case 0xcf:
+				key = sparse_keymap_entry_from_scancode(
+					priv->array, event);
+				input_report_key(priv->array, key->keycode, 0);
+				input_sync(priv->array);
+				return;
+		}
+	}
+
 	/* 0xC0 is for HID events, other values are for 5 button array */
 	if (event != 0xc0) {
 		if (!priv->array ||
-- 
2.20.1

